Change number of parents in iota message

muXxer opened the pull request:
# Description of change

This PR increases the number of parents of an IOTA message.
Parents must be at least 1, and a maximum of 8.
Parents must be unique and sorted by lexical order of their hashes.

The number was limited to two parents in former times, because tip selection was too slow and without the whiteflag approach, there was an attack vector that could render large parts of the tangle invalid. Since these problems are now gone, increasing the number of parents should increase the confirmation rate, mitigate blowball attacks and decrease the confirmation time.

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Tests have been modified.

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests using ginkgo that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

luca-moser commented on serializer.go:
```diff
@@ -158,13 +170,47 @@ func (s *Serializer) WriteVariableByteSlice(data []byte, lenType SeriSliceLength
+	data := producer()
+	sliceLength := len(data)
+
+	var arrayElementValidator ElementValidationFunc
+	if arrayRules != nil && deSeriMode.HasMode(DeSeriModePerformValidation) {
+		if err := arrayRules.CheckBounds(uint16(sliceLength)); err != nil {
+			s.err = errProducer(err)
+			return s
+		}
+
+		switch arrayRules.ValidationMode {
+		case ArrayValidModeNoValidation:
+		case ArrayValidModeDuplicates:
+			arrayElementValidator = arrayRules.ElementUniqueValidator()
+		case ArrayValidModeLexicalOrdering:
+			arrayElementValidator = arrayRules.LexicalOrderValidator()
+		case ArrayValidModeDuplicates | ArrayValidModeLexicalOrdering:
+			arrayElementValidator = arrayRules.LexicalOrderWithoutDupsValidator()
+		default:
+			panic(ErrUnknownArrayValidationMode)
+		}
+	}
```
This piece of code is duplicated at `Write64BytesArraySlice()`. Please make a separate function which given the validation mode inits the ElementValidationFunc appropriately. In fact, the switch statement should be within a method of `ArrayRules`.

luca-moser commented on serializer.go:
```diff
@@ -173,13 +219,47 @@ func (s *Serializer) Write32BytesArraySlice(data SliceOfArraysOf32Bytes, lenType
-		if _, err := s.buf.Write(data[i][:]); err != nil {
+		element := data[i][:]
+
+		// check array element validation against previous element
```
This comment is scattered in tests and here but it isn't always applicable. Maybe just remove it as it isn't correct.
For example, if only uniqueness is validated, the validator validates not just against the previous element. You know what I mean.

luca-moser commented on serializer.go:
```diff
@@ -478,7 +558,7 @@ func (d *Deserializer) readSliceLength(lenType SeriSliceLengthType, errProducer
 
 // ReadSliceOfArraysOf32Bytes reads a slice of arrays of 32 bytes.
-func (d *Deserializer) ReadSliceOfArraysOf32Bytes(slice *SliceOfArraysOf32Bytes, lenType SeriSliceLengthType, errProducer ErrProducer) *Deserializer {
+func (d *Deserializer) ReadSliceOfArraysOf32Bytes(f ReadSliceOfArraysOf32BytesConsumerFunc, deSeriMode DeSerializationMode, lenType SeriSliceLengthType, arrayRules *ArrayRules, errProducer ErrProducer) *Deserializer {
```
Same here, change this back to not take a consumer function.

luca-moser commented on serializer.go:
```diff
@@ -490,24 +570,53 @@ func (d *Deserializer) ReadSliceOfArraysOf32Bytes(slice *SliceOfArraysOf32Bytes,
 		return d
 	}
 
+	var arrayElementValidator ElementValidationFunc
+	if arrayRules != nil && deSeriMode.HasMode(DeSeriModePerformValidation) {
+		if err := arrayRules.CheckBounds(uint16(sliceLength)); err != nil {
+			d.err = errProducer(err)
+			return d
+		}
+
+		switch arrayRules.ValidationMode {
+		case ArrayValidModeNoValidation:
+		case ArrayValidModeDuplicates:
+			arrayElementValidator = arrayRules.ElementUniqueValidator()
+		case ArrayValidModeLexicalOrdering:
+			arrayElementValidator = arrayRules.LexicalOrderValidator()
+		case ArrayValidModeDuplicates | ArrayValidModeLexicalOrdering:
+			arrayElementValidator = arrayRules.LexicalOrderWithoutDupsValidator()
+		default:
+			panic(ErrUnknownArrayValidationMode)
+		}
+	}
+
```
Again, refactor this toa a method of `ArrayRules`.

luca-moser reviewed (changes requested):
The changes making the `MessageIDs` a slice of pointers to array pollutes the codebase with changes which otherwise wouldn't be necessary. Please revert `MessageIDs` back to be a slice of arrays. In any place where you need a slice or want to avoid copying the entire array, use `array[:]`.

muXxer commented:
I changed everything according your comments.
I added another function `RemoveDupsAndSortByLexicalOrderArrayOf32Bytes`.
Not sure if you like the name, because I don't :D
And I don't know where to put that function :shrug: 
Hope it fits.

luca-moser reviewed (approved):
Thanks for updating the PR, looks good now!
